### ISSUE TITLE
Add unified likelihood/prior function interface for improved computational efficiency and posterior repartitioning support

### ISSUE THREAD

**Issue by williamjameshandley** [handley-lab/blackjax#34]
## Issue Description

BlackJAX currently follows the PolyChordLite approach of using separate `logprior_fn` and `loglikelihood_fn` functions. However, MultiNest's unified interface approach offers significant advantages for computational efficiency, derived parameter computation, and advanced techniques like posterior repartitioning. This issue proposes adding support for a unified function interface while maintaining backward compatibility.

## Current BlackJAX Interface (PolyChordLite Style)

Currently in `blackjax/ns/nss.py`, nested sampling requires separate functions:

```python
def build_kernel(
    logprior_fn: Callable,           # Separate prior function
    loglikelihood_fn: Callable,     # Separate likelihood function
    num_inner_steps: int,
    # ... other parameters
):
```

This separation leads to:
- **Duplicated computations**: Shared calculations between prior and likelihood are performed twice
- **Inefficient memory usage**: Intermediate results cannot be reused
- **No derived parameters**: No natural way to compute derived quantities from shared calculations
- **Limited posterior repartitioning**: Advanced techniques require unified evaluation

## MultiNest Interface Pattern

Examining MultiNest examples (e.g., `/multinest/src/example_*/like.f90`), the interface uses a single function that:

1. **Transforms unit hypercube to physical parameters** (prior transformation)
2. **Computes likelihood with transformed parameters**  
3. **Supports derived parameters** calculated from shared intermediate results
4. **Returns both loglikelihood and transformed parameters**

### MultiNest Example Pattern
```fortran
subroutine getLogLike(Cube,n_dim,nPar,lnew,context)
    ! Input: Cube(nPar) contains unit hypercube coordinates
    ! Transform to physical parameters (prior)
    ! Compute likelihood using shared intermediate calculations
    ! Store derived parameters in Cube(n_dim+1:nPar) 
    ! Output: lnew = loglikelihood, Cube = physical + derived parameters
end subroutine
```

**Key insight**: Derived parameters are computed from the **same intermediate results** used for likelihood evaluation, avoiding expensive re-computation.

## Proposed Unified Interface

Add support for a unified function alongside existing separate functions:

```python
def build_kernel_unified(
    log_density_fn: Callable,       # New unified interface
    num_inner_steps: int,
    # ... other parameters  
) -> Callable:
    """Build NSS kernel with unified likelihood/prior function.
    
    Parameters
    ----------
    log_density_fn : Callable
        Function with signature: 
        (unit_params, context) -> (log_likelihood, physical_params, derived_params)
        
        - unit_params: Parameters in [0,1]^n unit hypercube
        - context: Optional additional data/settings
        - Returns: (log_likelihood, transformed_physical_params, derived_params)
    """
```

### Unified Function Interface
```python
def unified_density_fn(unit_params: ArrayTree, context: Optional[Dict] = None):
    """Unified likelihood/prior evaluation function with derived parameters.
    
    Parameters
    ----------  
    unit_params : ArrayTree
        Parameters in unit hypercube [0,1]^n
    context : Optional[Dict]
        Additional data (observations, hyperparameters, etc.)
        
    Returns
    -------
    tuple[float, ArrayTree, ArrayTree]
        - log_likelihood: Combined log(prior * likelihood)
        - physical_params: Transformed parameters in physical space
        - derived_params: Additional computed quantities from shared calculations
    """
    # Transform unit hypercube to physical parameters (prior transformation)
    physical_params = prior_transform(unit_params)
    
    # Compute expensive intermediate results ONCE
    intermediate_results = expensive_computation(physical_params, context)
    
    # Compute likelihood using intermediate results
    log_likelihood = compute_log_likelihood(intermediate_results, context)
    
    # Compute derived parameters from SAME intermediate results (no re-computation!)
    derived_params = compute_derived_quantities(intermediate_results, physical_params)
    
    return log_likelihood, physical_params, derived_params
```

### 1. **Computational Efficiency**
- **Shared calculations**: Prior transformation and likelihood can reuse intermediate results
- **Memory efficiency**: Single function call reduces memory allocation overhead
- **Vectorization**: Unified computation enables better JAX optimization

### 2. **Derived Parameters with Shared Calculations**
- **No re-computation**: Derived parameters use the same expensive intermediate calculations as likelihood
- **Natural integration**: MultiNest-style interface makes derived parameters a first-class feature
- **Flexible output**: Additional quantities for analysis computed efficiently within the same evaluation

### 3. **Posterior Repartitioning Support**
Posterior repartitioning requires unified evaluation for techniques like:
- **Bayesian Posterior Repartitioning (BPR)**: Redefines prior/likelihood while keeping their product constant
- **Annealing schedules**: Dynamically adjusts the balance between prior and likelihood
- **Hyperparameter inference**: Treats repartitioning parameters as part of the inference

### 4. **Advanced Sampling Techniques**  
- **Temperature annealing**: Unified interface supports tempered transitions
- **Hierarchical models**: Easier handling of hyperparameters and latent variables
- **Multi-fidelity sampling**: Different accuracy levels within single function

### 1. **Cosmological Parameter Estimation with Derived Parameters**
```python
def cosmology_unified(unit_params, context):
    # Transform to physical cosmological parameters 
    cosmo_params = transform_cosmology_params(unit_params)
    
    # Compute theory predictions ONCE (very expensive!)
    theory_cls = compute_cmb_power_spectra(cosmo_params)  
    
    # Compute likelihood using theory_cls
    log_likelihood = cmb_likelihood(theory_cls, context['cmb_data'])
    
    # Derive additional quantities from SAME theory calculation (no re-computation!)
    derived = {
        'H0': hubble_constant(cosmo_params),
        'sigma8': sigma8_from_cls(theory_cls),      # Reuses expensive theory_cls!
        'age_universe': cosmic_age(cosmo_params),   
        'matter_density': omega_m_from_cls(theory_cls),  # Also reuses theory_cls!
    }
    
    return log_likelihood, cosmo_params, derived

# Current approach would require:
# 1. compute_cmb_power_spectra() for likelihood
# 2. compute_cmb_power_spectra() AGAIN for sigma8 
# 3. compute_cmb_power_spectra() AGAIN for matter_density
# Unified approach: compute_cmb_power_spectra() only ONCE!
```

### 2. **Posterior Repartitioning Example**
```python  
def repartitioned_density(unit_params, context):
    beta = context.get('repartitioning_beta', 1.0)
    
    physical_params = prior_transform(unit_params)
    log_prior = compute_log_prior(physical_params)
    log_likelihood = compute_log_likelihood(physical_params, context)
    
    # Repartitioned density: prior^beta * likelihood  
    log_density = beta * log_prior + log_likelihood
    
    # Derived parameters include repartitioning diagnostics
    derived = {
        'beta': beta, 
        'log_prior': log_prior, 
        'log_likelihood': log_likelihood,
        'repartition_ratio': log_likelihood / log_prior
    }
    return log_density, physical_params, derived
```

### 3. **Hierarchical Model with Derived Individual-Level Parameters**
```python
def hierarchical_model(unit_params, context):
    # Split parameters into hyperparameters and individual parameters
    hyperparams = transform_hyperparams(unit_params[:n_hyper])
    individual_params = transform_individual_params(unit_params[n_hyper:], hyperparams)
    
    # Compute sufficient statistics ONCE (expensive for large datasets)
    sufficient_stats = compute_sufficient_statistics(context['data'])
    
    # Hierarchical likelihood using sufficient statistics
    log_likelihood = hierarchical_likelihood(individual_params, hyperparams, sufficient_stats)
    
    # Derived: individual-level summaries from SAME sufficient statistics
    derived = {
        'individual_means': individual_means_from_stats(sufficient_stats, individual_params),
        'individual_variances': individual_vars_from_stats(sufficient_stats, individual_params),
        'shrinkage_factors': compute_shrinkage(individual_params, hyperparams)
    }
    
    return log_likelihood, jnp.concatenate([hyperparams, individual_params]), derived
```

### 4. **Astrophysical Model with Shared Expensive Simulations**
```python
def stellar_evolution_model(unit_params, context):
    # Physical stellar parameters
    stellar_params = transform_stellar_params(unit_params)
    
    # Run expensive stellar evolution simulation ONCE
    evolutionary_track = run_stellar_evolution_simulation(stellar_params)  # Very expensive!
    
    # Likelihood from observations
    log_likelihood = compare_to_observations(evolutionary_track, context['observations'])
    
    # Many derived parameters from SAME evolutionary track (no re-simulation!)
    derived = {
        'main_sequence_lifetime': ms_lifetime_from_track(evolutionary_track),
        'final_mass': final_mass_from_track(evolutionary_track),
        'peak_luminosity': peak_luminosity_from_track(evolutionary_track),
        'helium_ignition_time': he_ignition_from_track(evolutionary_track),
        'surface_abundances': surface_abundances_from_track(evolutionary_track)
    }
    
    return log_likelihood, stellar_params, derived

# Current approach: run_stellar_evolution_simulation() called 6 times!
# Unified approach: run_stellar_evolution_simulation() called only ONCE!
```

### 1. **Backward Compatibility**
Maintain existing `build_kernel()` function with separate `logprior_fn`/`loglikelihood_fn`:

```python
def build_kernel(
    logprior_fn: Optional[Callable] = None,
    loglikelihood_fn: Optional[Callable] = None,
    log_density_fn: Optional[Callable] = None,  # New unified option
    # ... other parameters
):
    if log_density_fn is not None:
        # Use unified interface
        return build_unified_kernel(log_density_fn, ...)
    else:
        # Use existing separate functions interface
        return build_separate_kernel(logprior_fn, loglikelihood_fn, ...)
```

### 2. **Auto-detection**
Detect function signature to automatically choose appropriate interface:

```python
def auto_detect_interface(func):
    sig = inspect.signature(func)
    if len(sig.parameters) == 2:  # (unit_params, context)
        return 'unified'
    elif len(sig.parameters) == 1:  # (params)
        return 'separate'
    else:
        raise ValueError("Unsupported function signature")
```

### 3. **Bridge Functions**
Convert between interfaces when needed:

```python
def separate_to_unified(logprior_fn, loglikelihood_fn):
    def unified_fn(unit_params, context=None):
        # Simple transformation: assume uniform priors on unit hypercube
        physical_params = unit_params  # Could be more sophisticated
        log_prior = logprior_fn(physical_params)
        log_likelihood = loglikelihood_fn(physical_params)
        
        return log_prior + log_likelihood, physical_params, {}
    return unified_fn
```

### 4. **State Management for Derived Parameters**
Extend nested sampling state to handle derived parameters:

```python
@dataclass
class NSStateWithDerived:
    particles: ArrayTree          # Physical parameters
    log_likelihoods: Array        # Likelihood values
    log_priors: Array            # Prior values  
    derived_particles: ArrayTree  # Derived parameters (new!)
    # ... other state variables
```

### 1. **Nested Slice Sampling**
Update `blackjax/ns/nss.py` to support unified interface in the inner kernel and derived parameter storage.

### 2. **Base Nested Sampling**  
Extend `blackjax/ns/base.py` to handle derived parameters in state management and output.

### 3. **Output and Analysis**
Modify result structures to include derived parameters alongside standard output.

### 4. **Other Sampling Methods**
Consider extending to HMC, NUTS, and other methods that could benefit from unified interfaces and derived parameters.

## Related Research

This approach aligns with recent advances in nested sampling:
- **Posterior Repartitioning** [Handley & Speagle 2018, Speagle 2019]: Requires unified evaluation  
- **Dynamic Nested Sampling** [Higson et al. 2019]: Benefits from efficient likelihood evaluation
- **Multi-fidelity methods**: Natural fit with unified interface supporting different accuracy levels
- **Derived parameter analysis**: Standard in cosmological parameter estimation (e.g., Planck collaboration results)

## Related Issues

- **Issue #24**: Add derived parameters example to nested sampling documentation - this unified interface would make derived parameters a core feature
- **Issue #31**: Enhanced constraint parameters would work naturally with unified interface
- **Issues #20, #23, #30, #32, #33**: Improved robustness benefits from unified error handling

This enhancement would significantly improve BlackJAX's computational efficiency for expensive models, enable natural derived parameter computation, and support advanced nested sampling techniques while maintaining full backward compatibility.
